### ISSUE TITLE
Eager imports cause crash in partially built environment.

### ISSUE THREAD

**Issue by eli-schwartz** [serge-sans-paille/pythran#2194]
With pythran, numpy, and scipy installed on Gentoo Linux:

```
$ pythran --version
0.15.0
```

If I attempt to install numpy 2.x and scipy using the system package manager:
```
# unmask numpy and scipy release candidates

emerge -1 numpy scipy
```

numpy builds correctly. Then it tries to build scipy. However, since there is an older version of scipy installed, and scipy needs pythran to build, pythran fails to run.

Further investigation shows that pythran cannot even print its version:
```
/usr/lib/python3.11/importlib/__init__.py:126: UserWarning: A NumPy version >=1.22.4 and <1.29.0 is required for this version of SciPy (detected version 2.0.0rc1)
  return _bootstrap._gcd_import(name[level:], package, level)
Traceback (most recent call last):
  File "/usr/lib/python-exec/python3.11/pythran", line 5, in <module>
    from pythran.run import run
  File "/usr/lib/python3.11/site-packages/pythran/run.py", line 11, in <module>
    import pythran.types.tog
  File "/usr/lib/python3.11/site-packages/pythran/types/tog.py", line 10, in <module>
    from pythran.tables import MODULES, attributes
  File "/usr/lib/python3.11/site-packages/pythran/tables.py", line 4630, in <module>
    save_arguments((), MODULES)
  File "/usr/lib/python3.11/site-packages/pythran/tables.py", line 4576, in save_arguments
    save_arguments(module_name + (elem,), signature)
  File "/usr/lib/python3.11/site-packages/pythran/tables.py", line 4576, in save_arguments
    save_arguments(module_name + (elem,), signature)
  File "/usr/lib/python3.11/site-packages/pythran/tables.py", line 4580, in save_arguments
    themodule = import_module(".".join(module_name))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/scipy/special/__init__.py", line 777, in <module>
    from . import _ufuncs
  File "scipy/special/_ufuncs.pyx", line 1, in init scipy.special._ufuncs
ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject
```

This is awkward because if scipy was not installed at all, pythran would be perfectly okay. But because there is a broken scipy installed, it refuses to print --version. It is also hard to work around, because Gentoo is primarily a from-source distro. It is easy to mark scipy as "broken until it is rebuilt", but not easy to uninstall it from the live system in the middle of a rebuild.

It would be more convenient if there was some way pythran could avoid trying to import scipy until it knows it actually has to translate some scipy code, but I don't know how easy that is. Failing that, maybe it could catch the specific case of binary incompatibility and treat it the same as though scipy was uninstalled?

Either solution should permit completing the upgrade by successfully rebuilding scipy.

/cc @mgorny

**Comment by eli-schwartz** [comment]
Aside: it feels like pythran has some surprising slowness to print --version information on a cold start, which probably is a result of indirectly importing numpy and scipy at the top of the file which main() is in.

**Comment by serge-sans-paille** [comment]
> Aside: it feels like pythran has some surprising slowness to print --version information on a cold start, which probably is a result of indirectly importing numpy and scipy at the top of the file which main() is in.

This part was easy, fixed in #2195

**Comment by serge-sans-paille** [comment]
We're already trying to import `scipy` and `scipy.special` (see https://github.com/serge-sans-paille/pythran/blob/a1fc589aaa635597bef89311b4483900d2af30d6/pythran/tables.py#L4538).
From what I see, if we get a ValueError when trying to import `scipy.special` we should actually also dismiss `scipy`?

**Comment by eli-schwartz** [comment]
I think so, I just don't know enough about scipy internals to really know that ValueError is a "good enough" flag for that condition.

> This part was easy, fixed in #2195

Thanks! Moving exception definitions around brings back memories lol. Glad to see it was that easy though.
